server/src/main/java/org/opensearch/index/engine/IngestionEngine.java

```diff
@@ -0,0 +1,1016 @@
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.UnaryOperator;
+
+/**
+ * IngestionEngine is an engine that ingests data from a stream source.
+ */
+public class IngestionEngine extends Engine {
+
+    private volatile SegmentInfos lastCommittedSegmentInfos;
+    private final CompletionStatsCache completionStatsCache;
+    private final IndexWriter indexWriter;
+    private final OpenSearchReaderManager internalReaderManager;
+    private final ExternalReaderManager externalReaderManager;
+    private final Lock flushLock = new ReentrantLock();
+    private final ReentrantLock optimizeLock = new ReentrantLock();
+    private final OpenSearchConcurrentMergeScheduler mergeScheduler;
+    private final AtomicBoolean shouldPeriodicallyFlushAfterBigMerge = new AtomicBoolean(false);
+    private final TranslogManager translogManager;
+    private final DocumentMapperForType documentMapperForType;
+    private final IngestionConsumerFactory ingestionConsumerFactory;
+    protected StreamPoller streamPoller;
+
+    /**
+     * UUID value that is updated every time the engine is force merged.
+     */
+    @Nullable
+    private volatile String forceMergeUUID;
+
+    public IngestionEngine(EngineConfig engineConfig) {
+        super(engineConfig);
+        store.incRef();
+        boolean success = false;
+        try {
+            this.lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
+            this.completionStatsCache = new CompletionStatsCache(() -> acquireSearcher("completion_stats"));
+            IndexMetadata indexMetadata = engineConfig.getIndexSettings().getIndexMetadata();
+            assert indexMetadata != null;
+            mergeScheduler = new EngineMergeScheduler(engineConfig.getShardId(), engineConfig.getIndexSettings());
+            indexWriter = createWriter();
+            externalReaderManager = createReaderManager(new InternalEngine.RefreshWarmerListener(logger, isClosed, engineConfig));
+            internalReaderManager = externalReaderManager.internalReaderManager;
+            translogManager = new NoOpTranslogManager(
+                shardId,
+                readLock,
+                this::ensureOpen,
+                new TranslogStats(0, 0, 0, 0, 0),
+                new Translog.Snapshot() {
+                    @Override
+                    public void close() {}
+
+                    @Override
+                    public int totalOperations() {
+                        return 0;
+                    }
+
+                    @Override
+                    public Translog.Operation next() {
+                        return null;
+                    }
+                }
+            );
```

Comment:
Maybe create a static `EMPTY_TRANSLOG_SNAPSHOT` and reuse across this and `NoOpEngine`